### PR TITLE
Fix mount issues in containerized Kubelet

### PR DESCRIPTION
**Pull request by smarterclayton**
- nsenter was not located via path in the container
- nsenter needed '--' to separate command from args
- we need to search the parent FS at /rootfs in order to find mount/umount/findmnt

### CODE REVIEW

File: pkg/util/mount/nsenter_mount.go [hunk 1]
```diff
@@ -39,14 +40,31 @@ import (
...
 //     bind-mount in this docker version.
 // 2.  The host's root filesystem must be available at /rootfs
-// 3.  The nsenter binary must be at /nsenter in the container's filesystem.
+// 3.  The nsenter binary must be on the PATH in the container's filesystem.
 // 4.  The Kubelet process must have CAP_SYS_ADMIN (required by nsenter); at
 //     the present, this effectively means that the kubelet is running in a
 //     privileged container.
+// 5.  The volume path used by the Kubelet must be the same inside and outside
+//     the container and be writable by the container (to initialize volume)
+//     contents. TODO: remove this requirement.
```
Comment (kelseyhightower) [inline pkg/util/mount/nsenter_mount.go]: Should we document the fact that the container image must contain mount, umount, and findmnt, or the rootfs you mount inside the container does?

Comment (smarterclayton) [inline pkg/util/mount/nsenter_mount.go]: We should

File: pkg/util/mount/nsenter_mount.go [hunk 2]
```diff
@@ -39,14 +40,31 @@ import (
 //     performed in the host's mount namespace do not propagate out to the
 //     bind-mount in this docker version.
 // 2.  The host's root filesystem must be available at /rootfs
-// 3.  The nsenter binary must be at /nsenter in the container's filesystem.
+// 3.  The nsenter binary must be on the PATH in the container's filesystem.
```
Comment (lavalamp) [inline pkg/util/mount/nsenter_mount.go]: Whose PATH?

File: pkg/util/mount/nsenter_mount.go [hunk 3]
```diff
@@ -39,14 +40,31 @@ import (
...
 // For more information about mount propagation modes, see:
 //   https://www.kernel.org/doc/Documentation/filesystems/sharedsubtree.txt
-type NsenterMounter struct{}
+type NsenterMounter struct {
+	// the path on the parent filesystem that contains mount, umount, and findmnt
+	hostBinPath string
+}
+
+func NewNsenterMounter() *NsenterMounter {
+	// search for the mount command in other locations besides /usr/bin
```
Comment (lavalamp) [inline pkg/util/mount/nsenter_mount.go]: Can you add a comment with justification for expecting all of the binaries to be located in the same directory? (i.e., why not store three path variables, one for each?)

File: pkg/util/mount/nsenter_mount_unsupported.go [hunk 4]
```diff
@@ -20,6 +20,10 @@ package mount
 
 type NsenterMounter struct{}
 
+func NewNsenterMounter() *NsenterMounter {
+	return &NsenterMounter{}
```
Comment (lavalamp) [inline pkg/util/mount/nsenter_mount_unsupported.go]: Shouldn't you add "/usr/bin" as a default hostBinPath?

**Comment by lavalamp** [comment]
Minor comments.

File: pkg/util/mount/nsenter_mount.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/util/mount/nsenter_mount.go]: Good point

File: pkg/util/mount/nsenter_mount_unsupported.go (hunk 4 above)
Comment (smarterclayton) [inline pkg/util/mount/nsenter_mount_unsupported.go]: This is for unsupported and it doesn't call those methods.

**Comment by smarterclayton** [comment]
Comments addressed - I search for each binary now in a list of paths, first path wins.  If doesn't exist will default to `/<binary>` and the other mount commands will fail and log

**Comment by lavalamp** [comment]
Code LGTM. @dchen1107 to evaluate for 1.0 & second LGTM.

**Comment by kelseyhightower** [comment]
LGTM

**Comment by dchen1107** [comment]
@smarterclayton Can we hold this one for a few weeks, right after 1.0 release?

**Comment by smarterclayton** [comment]
Sure.

**Comment by pmorie** [comment]
@dchen1107 should this have the 1.0-post label? I have been working on mount propagation issues in docker and had to aply this patch to test correctly.  Just want to make sure this goes in ASAP since I am going to be working in this area.

**Comment by smarterclayton** [comment]
Docker 1.7 is broken in a container (except on Red Hat distros) because mount propagation is PRIVATE, not SHARED.  Mount-prop fix is being tracked in libcontainer, will probably require docker changes as well.  The other option may be to start with the hosts mount namespace.

**Comment by smarterclayton** [comment]
@dchen1107 second review?

**Comment by lvlv** [comment]
My centos 6.5 + docker 1.7.1 + hyperkube + Kubernetes v1.0.1 is suffering this issue. Is it possible to at least merge "nsenter needed '--' to separate command from args" to release branch?

Here is the log:
I0731 07:52:48.829233       1 nsenter_mount.go:79] nsenter Mounting tmpfs /var/lib/kubelet/pods/a9d8d942-3729-11e5-abda-080027ce083d/volumes/kubernetes.io~secret/dns-token tmpfs []
I0731 07:52:48.829242       1 nsenter_mount.go:82] Mount command: /nsenter [--mount=/rootfs/proc/1/ns/mnt /usr/bin/mount -t tmpfs -o  tmpfs /var/lib/kubelet/pods/a9d8d942-3729-11e5-abda-080027ce083d/volumes/kubernetes.io~secret/dns-token]
I0731 07:52:48.830048       1 nsenter_mount.go:86] Output from mount command: nsenter: failed to parse pid: 'tmpfs'
E0731 07:52:48.830076       1 kubelet.go:1225] Unable to mount volumes for pod "kube-dns-fqgan_default": exit status 1; skipping pod

**Comment by smarterclayton** [comment]
I think we already got 1 LGTM, just needs a sign off from Dawn if she wants
to review it.

On Jul 31, 2015, at 5:42 AM, Lv Lv <notifications@github.com> wrote:

My centos 6.5 + docker 1.7.1 + hyperkube + Kubernetes v1.0.1 is suffering
this issue. Is it possible to at least merge "nsenter needed '--' to
separate command from args" to release branch?

Here is the log:
I0731 07:52:48.829233 1 nsenter_mount.go:79] nsenter Mounting tmpfs
/var/lib/kubelet/pods/a9d8d942-3729-11e5-abda-080027ce083d/volumes/
kubernetes.io~secret/dns-token tmpfs []
I0731 07:52:48.829242 1 nsenter_mount.go:82] Mount command: /nsenter
[--mount=/rootfs/proc/1/ns/mnt /usr/bin/mount -t tmpfs -o tmpfs
/var/lib/kubelet/pods/a9d8d942-3729-11e5-abda-080027ce083d/volumes/
kubernetes.io~secret/dns-token]
I0731 07:52:48.830048 1 nsenter_mount.go:86] Output from mount command:
nsenter: failed to parse pid: 'tmpfs'
E0731 07:52:48.830076 1 kubelet.go:1225] Unable to mount volumes for pod
"kube-dns-fqgan_default": exit status 1; skipping pod

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/9976#issuecomment-126624808>
.

**Comment by timothysc** [comment]
This is very useful for automating scale out setups. 

/cc @wojtek-t @rrati @jbeda

**Comment by dchen1107** [comment]
Sorry for delaying the review for this PR. LGTM

**Pull request merged**